File: core-extra/arazzo-parser/src/main/java/com/webfuzzing/arazzo/resolver/ArazzoReferenceResolver.java
package com.webfuzzing.arazzo.resolver;

import com.fasterxml.jackson.databind.JsonNode;
import com.webfuzzing.arazzo.models.domain.*;
import io.swagger.v3.core.util.Json;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.media.Schema;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Resolver class responsible for resolving Arazzo references.
 */
public class ArazzoReferenceResolver {
    private final String COMPONENTS_PREFIX = "$components.";
    private final String SUCCESS_ACTIONS = "successActions";
    private final String FAILURE_ACTIONS = "failureActions";
    private final String PARAMETERS = "parameters";
    private final String OPENAPI_SCHEMAS_PREFIX = "/components/schemas/";

    private Components components;
    private JsonNode arazzoJsonNode;
    private OpenAPI openApi;

    public ArazzoReferenceResolver(Components components, JsonNode arazzoJsonNode, OpenAPI openAPI) {
        this.components = components;
        this.arazzoJsonNode = arazzoJsonNode;
        this.openApi = openAPI;
    }

    public void setComponents(Components components) {
        this.components = components;
    }

    public void setArazzoJsonNode(JsonNode arazzoJsonNode) {
        this.arazzoJsonNode = arazzoJsonNode;
    }

    public void setOpenApi(OpenAPI openApi) {
        this.openApi = openApi;
    }

    /**
     * Resolve the {@link SuccessReusable} references to get a complete list of {@link SuccessAction}
     */
    public List<SuccessAction> resolveSuccessReusable(List<SuccessReusable> items) {
        if (items == null) {
            return null;
        }

        return items.stream().map(item -> {
            if (item instanceof SuccessReusable.Success) {
                SuccessReusable.Success action = (SuccessReusable.Success) item;
                return action.getAction();
            }
            SuccessReusable.ReusableObj reusableObj = (SuccessReusable.ReusableObj) item;
            return (SuccessAction) resolveReusableWithPrefix(reusableObj.getReusable(), this.SUCCESS_ACTIONS);
        }).collect(Collectors.toList());
    }

    /**
     * Resolve the {@link FailureReusable} references to get a complete list of {@link FailureAction}
     */
    public List<FailureAction> resolveFailureReusable(List<FailureReusable> items) {
        if (items == null) {
            return null;
        }

        return items.stream().map(item -> {
            if (item instanceof FailureReusable.Failure) {
                FailureReusable.Failure action = (FailureReusable.Failure) item;
                return action.getAction();
            }
            FailureReusable.ReusableObj reusableObj = (FailureReusable.ReusableObj) item;
            return (FailureAction) resolveReusableWithPrefix(reusableObj.getReusable(), this.FAILURE_ACTIONS);
        }).collect(Collectors.toList());
    }

    /**
     * Resolve the {@link ParameterReusable} references to get a complete list of {@link Parameter}
     */
    public List<Parameter> resolveParametersReusable(List<ParameterReusable> items) {
        if (items == null) {
            return null;
        }

        return items.stream().map(item -> {
            if (item instanceof ParameterReusable.Param) {
                ParameterReusable.Param param = (ParameterReusable.Param) item;
                return param.getParameter();
            }
            ParameterReusable.ReusableObj reusableObj = (ParameterReusable.ReusableObj) item;
            return (Parameter) resolveReusableWithPrefix(reusableObj.getReusable(), this.PARAMETERS);
        }).collect(Collectors.toList());
    }

    private Object resolveReusableWithPrefix(Reusable reusable, String expectedPrefix) {
        if (components == null) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Can't reference with no Components");
        }

        String reference = reusable.getReference();
        if (!reference.startsWith(this.COMPONENTS_PREFIX)) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Invalid reference (" + reference + "). Expected to point to '" + this.COMPONENTS_PREFIX + "'");
        }

        String referenceWithoutComponents = reference.substring(this.COMPONENTS_PREFIX.length());
        if (!referenceWithoutComponents.startsWith(expectedPrefix)) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Invalid reference (" + referenceWithoutComponents + "). Expected to point to '" + expectedPrefix +"'");
        }

        String referenceClean = referenceWithoutComponents.substring((expectedPrefix + ".").length());
        Object resolve;
        switch (expectedPrefix) {
            case SUCCESS_ACTIONS:
                resolve = (components.getSuccessAction() != null) ? components.getSuccessAction().get(referenceClean) : null;
                break;
            case FAILURE_ACTIONS:
                resolve = (components.getFailureAction() != null) ? components.getFailureAction().get(referenceClean) : null;
                break;
            case PARAMETERS:
                resolve = (components.getParameters() != null) ? components.getParameters().get(referenceClean) : null;
                break;
            default:
                resolve = null;
                break;
        }

        if (resolve == null) {
            throw new IllegalArgumentException("Arazzo Parsing Error: The " + expectedPrefix + ": '" + referenceClean + "' is not in the components.");
        }

        return resolve;

    }

    public Schema<?> resolveJsonPointer(String reference) {
        if (reference.startsWith("#/")) {
            return resolveJsonPointerLocal(reference);
        }
        return resolveJsonPointerExternal(reference);
    }

    private Schema<?> resolveJsonPointerLocal(String reference) {
        if (arazzoJsonNode == null) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Can't reference with no Arazzo Document");
        }

        String jsonPointer = reference.substring(1);

        JsonNode result = arazzoJsonNode.at(jsonPointer);

        if (result.isMissingNode()) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Can't reference '" + reference + "'");
        }

        return Json.mapper().convertValue(result, Schema.class);
    }

    private Schema<?> resolveJsonPointerExternal(String reference) {
        if (openApi == null) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Can't reference with no OpenApi Document");
        }

        String[] tokens = reference.split("#");
        if (tokens.length < 2) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Error reference (" + reference + "). '#' Is mandatory");
        }

        String jsonPointer = tokens[1];
        if (!jsonPointer.startsWith(this.OPENAPI_SCHEMAS_PREFIX)) {
            throw new IllegalArgumentException("Arazzo Parsing Error: Error reference (" + reference + "). \"" + this.OPENAPI_SCHEMAS_PREFIX + "\" Is mandatory for references to OpenApi");
        }

        String schemaName = jsonPointer.substring(this.OPENAPI_SCHEMAS_PREFIX.length());
        Schema<?> result = null;
        if (openApi.getComponents() != null && openApi.getComponents().getSchemas() != null) {
            result = openApi.getComponents().getSchemas().get(schemaName);
        }

        if (result == null) {
            throw new IllegalArgumentException("Arazzo Parsing Error: (" + reference + ") reference does not exist in the OpenApi document");
        }

        return result;
    }

}

File: core/src/main/kotlin/org/evomaster/core/problem/rest/arazzo/resolver/ArazzoReferenceResolver.kt
package org.evomaster.core.problem.rest.arazzo.resolver

import com.fasterxml.jackson.databind.JsonNode
import io.swagger.v3.core.util.Json
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.media.Schema
import org.evomaster.core.problem.rest.arazzo.models.*

class ArazzoReferenceResolver(
    val components: Components?,
    val arazzoJsonNode: JsonNode?,
    val openApi: OpenAPI?
) {

    fun resolveSuccessReusable(items: List<SuccessReusable>?): List<SuccessAction>? {
        if (items == null) return null

        return items.map { item ->
            when (item) {
                is SuccessReusable.Success -> item.action
                is SuccessReusable.ReusableObj -> resolveReusableWithPrefix(item.reusable, "successActions") as SuccessAction
            }
        }
    }

    fun resolveFailureReusable(items: List<FailureReusable>?): List<FailureAction>? {
        if (items == null) return null

        return items.map { item ->
            when (item) {
                is FailureReusable.Failure -> item.action
                is FailureReusable.ReusableObj -> resolveReusableWithPrefix(item.reusable, "failureActions") as FailureAction
            }
        }
    }

    fun resolveParametersReusable(items: List<ParameterReusable>?): List<Parameter>? {
        if (items == null) return null

        return items.map { item ->
            when (item) {
                is ParameterReusable.Param -> item.parameter
                is ParameterReusable.ReusableObj -> resolveReusableWithPrefix(item.reusable, "parameters") as Parameter
            }
        }
    }

    private fun resolveReusableWithPrefix(reusable: Reusable, prefixExpected: String) : Any {
        if (components == null) {
            throw IllegalArgumentException("Arazzo Parsing Error: Can't reference with no Components")
        }

        val reference = reusable.reference

        if (reference !is RuntimeExpression.Components) {
            throw IllegalArgumentException(
                "Arazzo Parsing Error: A reference to Components was expected."
            )
        }

        if (!reference.name.startsWith(prefixExpected)) {
            throw IllegalArgumentException(
                "Arazzo Parsing Error: Invalid reference (${reference.name}). Expected to point to '${prefixExpected}'"
            )
        }

        val actionName = reference.name.removePrefix(prefixExpected + ".")
        val resolve = when(prefixExpected) {
            "successActions" -> components.successActions?.get(actionName)
            "failureActions" -> components.failureActions?.get(actionName)
            "parameters" -> components.parameters?.get(actionName)
            else -> null
        }

        if (resolve == null) {
            throw IllegalArgumentException(
                "Arazzo Parsing Error: The ${prefixExpected}: '$actionName' is not in the components."
            )
        }

        return resolve
    }

    private fun resolveJsonPointer(reference: String) : Schema<*>? {
        if (reference.startsWith("#/")) {
            return resolveJsonPointerLocal(reference)
        }
        return resolveJsonPointerExternal(reference)
    }

    private fun resolveJsonPointerLocal(reference: String) : Schema<*>? {
        if (arazzoJsonNode == null) {
            throw IllegalArgumentException("Arazzo Parsing Error: Can't reference with no Arazzo Document")
        }

        val jsonPointer = reference.substring(1)

        val result = arazzoJsonNode.at(jsonPointer)

        if (result.isMissingNode) {
            throw IllegalArgumentException("Arazzo Parsing Error: Can't reference '${reference}'")
        }

        return Json.mapper().convertValue(result, Schema::class.java)
    }

    private fun resolveJsonPointerExternal(reference: String) : Schema<*> {
        if (openApi == null) {
            throw IllegalArgumentException("Arazzo Parsing Error: Can't reference with no OpenApi Document")
        }

        val tokens = reference.split("#")
        if (tokens.size < 2) {
            throw IllegalArgumentException("Arazzo Parsing Error: Error reference (${reference}). '#' Is mandatory")
        }

        val jsonPointer = tokens[1]
        val expectedPrefix = "/components/schemas/"
        if (!jsonPointer.startsWith(expectedPrefix)) {
            throw IllegalArgumentException("Arazzo Parsing Error: Error reference (${reference}). \"/components/schemas/\" Is mandatory for references to OpenApi")
        }

        val schemaName = jsonPointer.removePrefix(expectedPrefix)
        val result = openApi.components?.schemas?.get(schemaName)
            ?: throw IllegalArgumentException("Arazzo Parsing Error: (${reference}) reference does not exist in the OpenApi document")

        return result
    }

}